### ISSUE TITLE
Too many unnecessary copies

### ISSUE THREAD

**Issue by dev-ardi** [veeso/ssh2-config#6]
## Description

Usage of `String` and `Vec` where `&str` and `impl Iterator` would suffice.

Example:

```rs
    fn parse_string(args: Vec<String>) -> SshParserResult<String> {
        if let Some(s) = args.get(0) {
            Ok(s.to_string())
        } else {
            Err(SshParserError::MissingArgument)
        }
    }
```
would be the same as 
```rs
    fn parse_str<'a>(mut args: impl Iterator<Item = &'a str>) -> SshParserResult<&'a str> {
        if let Some(s) = args.next() {
            Ok(s)
        } else {
            Err(SshParserError::MissingArgument)
        }
    }
```
Without copying any data.

**Comment by veeso** [comment]
fixed in 0.2.1

**Comment by dev-ardi** [comment]
Thanks for fixing that example but I think this should be left open as there are still many unnecessary copies, and get fixed over time.

I was working on it yesterday, a zero-copy version where everything was `&'a str` where `'a` was the lifetime of the ssh config itself but I kid you not, I deleted ALL of the changes before comitting them 😢
What a way to waste an afternoon, now I'm not motivated anymore haha sorry.

If you want more details and would like to try (it can be incrementally adopted) let me know so I can help.
